### PR TITLE
Improved data fetching process

### PR DESCRIPTION
**Pull request by tgelo**
Thread management logic moved to data fetchers beans. 
Now, manually and scheduled fetching process blocks each other.

### CODE REVIEW

File: src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java [hunk 1]
```diff
@@ -58,9 +44,8 @@ private AjaxLink createFetchBlogPostsDataButton(CustomFeedbackPanel feedback) {
             @Override
             public void onClick(AjaxRequestTarget target) {
                 log.debug("Fetching blog posts data clicked");
-                if (fetchingBlogPostExecutor.getActiveCount() == 0) {
-                    fetchingBlogPostExecutor
-                        .submit(blogPostsFetchingScheduler::checkRssForNewBlogPosts);
+                if (!blogPostsFetcher.isFetchingProcessInProgress()) {
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java]: Same `if` statement is already used in `refreshPosts()` can we remove it from here and make `isFetchingProcessInProgress` private?

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java [hunk 2]
```diff
@@ -7,35 +7,74 @@
...
             .forEach(person -> rssCheckingActor.tell(new RssLink(person), ActorRef.noSender()));
+
+        final Metadata dateOfLastFetch = metadataRepository
+            .findByName(MetadataKeys.DATE_OF_LAST_FETCHING_BLOG_POSTS);
+        dateOfLastFetch.setValue(nowProvider.now().toString());
+        metadataRepository.save(dateOfLastFetch);
+    }
+
+    public boolean isFetchingProcessInProgress() {
+        return fetchingBlogPostExecutor.getActiveCount() != 0;
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: I know that this is just moved from other class, but I'm wondering if this check is reliable. According to javadoc:

> Returns the **approximate** number of threads that are actively executing tasks.

Shouldn't we track `Feature` returned from latest call to `submit` and check whether it `isDone` or not?

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java [hunk 3]
```diff
@@ -7,35 +7,74 @@
...
     public void refreshPosts() {
+        if (isFetchingProcessInProgress()) {
+            log.info("Fetching blog posts data already in progress");
+            return;
+        }
+        fetchingBlogPostExecutor
+            .submit(this::startFetchingProcess);
+    }
+
+    public void startFetchingProcess() {
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: I guess this method can be private.

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java [hunk 4]
```diff
@@ -7,35 +7,74 @@
...
         blogPostStoringActor =
             actorSystem.actorOf(NewBlogPostStoringActor.props(blogPostRepository));
         rssCheckingActor = actorSystem.actorOf(new RoundRobinPool(10)
             .props(RssCheckingActor.props(blogPostStoringActor, syndFeedFactory)), "rss-checkers");
+        this.metadataRepository = metadataRepository;
+        this.nowProvider = nowProvider;
     }
 
     public void refreshPosts() {
+        if (isFetchingProcessInProgress()) {
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: I think we are missing a test case here.

File: src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java (hunk 1 above)
Comment (tgelo) [inline src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java]: This statement is needed to send correct information to the user.

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java (hunk 3 above)
Comment (tgelo) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: True!

File: src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java (hunk 1 above)
Comment (airborn) [inline src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java]: Maybe `refreshPost` should return some kind of status marking request as `ACCEPTED` or `REJECTED`. Maybe even with appropriate message included.

Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/admin_panel/moderation/FetchingDataPanel.java]: I agree, if `refreshPost` returns some request state then this method `isFetchingProcessInProgress` could me moved inside fetcher.

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java (hunk 2 above)
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: But generally, to fetch blogs we create X messages and send them to N Akka actors to checke these rss feeds. They consume these X messages asynchronously so I am not sure that this `ThreadPoolExecutor` is helping us to check if we are "in progress" of anything :)

Maybe, after rethinking, we can ignore this problem as not important enough as firing two fetching processes won't do any harm (no duplicates, only more traffic)

File: src/main/java/com/jvm_bloggers/core/data_fetching/blogs/BloggersDataFetcher.java [hunk 5]
```diff
@@ -57,12 +88,15 @@ private void refreshBloggersDataFor(Optional<URL> blogsDataUrl, BlogType blogTyp
                 BloggersData bloggers = mapper.readValue(blogsDataUrl.get(), BloggersData.class);
                 bloggers.getBloggers().stream().forEach(it -> it.setBlogType(blogType));
                 bloggersDataUpdater.updateData(bloggers);
-            } catch (IOException exception) {
+            } catch (Exception exception) {
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blogs/BloggersDataFetcher.java]: Why we are widening the type of the caught exception here? Isn't it a bad code smell to catch general `Exception`?

Comment (tgelo) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blogs/BloggersDataFetcher.java]: My oversight.

File: src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java (hunk 2 above)
Comment (tgelo) [inline src/main/java/com/jvm_bloggers/core/data_fetching/blog_posts/BlogPostsFetcher.java]: In this case I agree with you. But this logic is also used with `BloggersDataFetcher.refreshData() `which process is synchronous. Therefore, while designing this solution I chose `ThreadPoolExecutor`. 

I see two solutions:

1) Change `BloggersDataFetcher.refreshData()` process to asynchronous using (akka actor?). Then, abandon `ThreadPoolExecutor` and `isFetchingProcessInProgress()`.
I dont know akka, so If It is possible, I will be happy while learning akka.

2) Develop two different solutions for `BloggersDataFetcher` (with new thread and check If process is already running) and `BlogPostsFetcher` (without new thread).

**Comment by tgelo** [comment]
Fetching bloggers data process is using now akka actors.  Thread pool executors were deleted.

**Comment by tdziurko** [comment]
Merged into master in another PR.

**Pull request closed without merging**